Running an unstaked Access Node

vishalchangrani opened the pull request:
This is the first set of changes needed to run an Unstaked Access node (an access node whose identity is not part of the identity table). This brings up the AN as a stand alone program or as part of the localnet.

There are two different networks - unstaked network and staked network. An unstaked access node only joins the unstaked network. A staked access node joins both the unstaked and the staked network.

![UnstakedAN_(1)](https://user-images.githubusercontent.com/1117327/124840566-22d1ae00-df40-11eb-87d3-fe5e28c1e95f.jpg)


This PR introduces the following changes:
1. Changes to `scaffold.go` to allow initializing an unstaked access node a little differently than a regular node via new `PreInit` functions and exported public functions.
2. Adding `AccessNodeBuilder` to initialize and run a staked or an unstaked access node. 
2. Changes to localnet to generate an unstaked access node along with other nodes. The keys for the unstaked AN are generated like any other node but the root snapshot excludes the unstaked node.

Currently, the unstaked Access node does expect the genesis data (root snapshot) AND the `node-info.json`. It reads the `nodeid` and the networking key from it (not the staking key).

I could _generate_ the networking key and the node id _on the fly_ as the node starts but I am wondering if there is value in asking the node operator to provide a node id (maybe for auditing traffic coming into a staked access node).

the unstaked access node can be run using the args:

> --nodeid=a0e1d65a9a63accda037691321df0de0605e36577ce1eba6683ace6751ea9aea
--bootstrapdir=/Users/vishalchangrani/go/src/github.com/onflow/flow-go/integration/localnet/bootstrap 
--datadir=/Users/vishalchangrani/go/src/github.com/onflow/flow-go/integration/localnet/data
--loglevel=DEBUG
**--staked=false** 
--unstaked-bind-addr=localhost:9658
--staked-access-node-id eb5baa26e18f98f90769c844bd2414a465dc6f8e50f33d73985d65b0c03b0aa8

the staked access node can be run using the args:

> --nodeid=eb5baa26e18f98f90769c844bd2414a465dc6f8e50f33d73985d65b0c03b0aa8 
--bootstrapdir=/Users/vishalchangrani/go/src/github.com/onflow/flow-go/integration/localnet/bootstrap 
--datadir=/Users/vishalchangrani/go/src/github.com/onflow/flow-go/integration/localnet/data 
--loglevel=DEBUG 
**--staked=true**
--unstaked-bind-addr=localhost:9658
--bind=localhost:8569


To run in IDE specify params here -
![Screen Shot 2021-06-29 at 3 19 03 PM](https://user-images.githubusercontent.com/1117327/123874667-5f245f00-d8ed-11eb-9d16-45255485c211.png)

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -135,6 +135,7 @@ type FlowNodeBuilder struct {
 	postInitFns       []func(*FlowNodeBuilder)
 	stakingKey        crypto.PrivateKey
 	networkKey        crypto.PrivateKey
+	unstaked          bool
```
this flag is used to determine if we are running a staked or an unstaked access node.
I had to be here in scaffold.go and not in the `cmd/access/main.go` file because the `initNodeInfo` happens here which initializes the node info (node id, keys, address etc.).

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -526,8 +532,20 @@ func (fnb *FlowNodeBuilder) initState() {
-	fnb.MustNot(err).Msgf("node identity not found in the identity list of the finalized state: %v", myID)
+
+	var self *flow.Identity
+	if fnb.unstaked {
```
In the case of the unstaked node, no attempts to check if the the node exist in the protocol data should be made.

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -526,8 +532,20 @@ func (fnb *FlowNodeBuilder) initState() {
+	}
 
 	// Verify that my role (as given in the configuration) is consistent with the protocol state.
 	// We enforce this strictly for MainNet. For other networks (e.g. TestNet or BenchNet), we
```
this existing check, will also catch the error if the `Role` in the node-info.json is anything other than `access`

vishalchangrani commented on integration/localnet/Makefile:
```diff
@@ -48,3 +48,19 @@ stop:
 load:
 	go run --tags relic ../loader -log-level info -tps 1,0,10,0,100 -tps-durations 30s,10s,30s,10s
+
+unstaked-an:
```
extra target to bring up a setup with 1 unstaked AN, 1 staked AN and the other nodes. This should help local testing.

vishalchangrani commented on integration/localnet/bootstrap.go:
```diff
@@ -124,6 +128,15 @@ func main() {
 		panic(err)
 	}
 
+	fmt.Println("Node bootstrapping data generated...")
```
to help distinguish between the staked versus the unstaked AN.

vishalchangrani replied:
e.g. output

> Node bootstrapping data generated...
> 0: execution-af7528fdf3b6c7310e711bf388830d03adadd198eb9177294cfcc8e1d7fc8497@execution_1:2137=1000
> 1: collection-b0fecccd452b357bf4a3c59d61577985ae9471c6dc053f88c2617067692d7a11@collection_1:2137=1000
> 2: consensus-7ddc00b3abb883f57aecf2f4ce4bf8c20ac0f82a54b12337d3c3e4b52f485a5f@consensus_1:2137=1000
> 3: verification-b2aab0841ba43c0f6a9052acce534707faea115a8b153858700ba91b5b7f122f@verification_1:2137=1000
> 4: access-6e8be2562605a2259af952329f23dce62a07222458c1fa07e9201d0537c87549@access_1:2137=1000
> 5: access-4a75b3786b48c920921b83cbf642c005c9d34aa23a2d2724daf99c1422236814@access_2:2137=1000 (unstaked)
> Bootstrapping success!

vishalchangrani commented on integration/testnet/network.go:
```diff
@@ -583,19 +584,43 @@ func BootstrapNetwork(networkConf NetworkConfig, bootstrapDir string) (*flow.Blo
 		return nil, nil, nil, nil, fmt.Errorf("failed to setup keys: %w", err)
 	}
 
+	// write private key files for each node
```
moved up to separate out staked vs non-staked nodes before calling `runDKG` for only the staked nodes.

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -204,6 +206,16 @@ func (fnb *FlowNodeBuilder) enqueueNetworkInit() {
 			return nil, fmt.Errorf("could not generate libp2p node factory: %w", err)
 		}
 
+		if fnb.unstaked {
```
message validators are a sequence of validators that each incoming message passes through. Currently, there are only two type of validators - SenderValidator and TargetValidator. SendorValidator validates that the message received was not sent my the same node and TargetValidator validates that the targetIDs in the message include this node. For the unstaked AN, we do not want TargetValidator since the Staked AN will not be specifying any specific target node ids but will be instead just broadcsating it to all.

vishalchangrani commented on network/p2p/middleware.go:
```diff
@@ -114,7 +114,7 @@ func NewMiddleware(log zerolog.Logger,
 }
 
-func defaultValidators(log zerolog.Logger, flowID flow.Identifier) []network.MessageValidator {
+func DefaultValidators(log zerolog.Logger, flowID flow.Identifier) []network.MessageValidator {
```
Making public so that it can be used in scaffold.

vishalchangrani commented on cmd/access/main.go:
```diff
@@ -75,10 +75,11 @@ func main() {
 
 	cmd.FlowNode(flow.RoleAccess.String()).
-		ExtraFlags(func(flags *pflag.FlagSet) {
+		ExtraFlagsWithFnb(func(flags *pflag.FlagSet, fnb *cmd.FlowNodeBuilder) {
```
The `unstaked` command line arg determines if the AN starts as a staked (unstaked=false) or an unstaked tunstaked=rue) node.
The `staked-access-node-addr` is the address of the staked node this unstaked access node is tethered to.
While the `unstaked` and `staked-access-node-addr` are access node only params, the FlowNodeBuilder initialization needs to be changed for the unstaked node. Hence,  the `staked` variable is defined within scaffold.go. `ExtraFlagsWithFnb` is a new method which allows changing the fnb.Staked flag when parsing the command line arg.

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -135,6 +136,7 @@ type FlowNodeBuilder struct {
 	postInitFns       []func(*FlowNodeBuilder)
 	stakingKey        crypto.PrivateKey
 	networkKey        crypto.PrivateKey
+	Unstaked          bool
```
This is the flag used to determine if the node should run as staked (`Unstaked=false`) or unstaked (`Unstaked=true`). Although it is only used for an access node, it is define here in scaffold because it needs to influence how the node is initialized by the flow node builder.

vishalchangrani commented on cmd/access/main.go:
```diff
@@ -75,35 +75,38 @@ func main() {
-			flags.StringToIntVar(&apiBurstlimits, "api-burst-limits", nil, "burst limits for Access API methods e.g. Ping=100,GetTransaction=100 etc.")
-		}).
+	fnb := cmd.FlowNode(flow.RoleAccess.String())
+	fnb.ExtraFlags(func(flags *pflag.FlagSet) {
```
Two new params -
1. `unstaked` - true if this node should run as an unstaked AN.
2. `staked-access-node-addr` - the address of the staked AN this unstaked AN is tethered to.

smnzhu commented on cmd/access/main.go:
```diff
@@ -75,35 +75,38 @@ func main() {
+		flags.StringVarP(&nodeInfoFile, "node-info-file", "", "", "full path to a json file which provides more details about nodes when reporting its reachability metrics")
+		flags.StringToIntVar(&apiRatelimits, "api-rate-limits", nil, "per second rate limits for Access API methods e.g. Ping=300,GetTransaction=500 etc.")
+		flags.StringToIntVar(&apiBurstlimits, "api-burst-limits", nil, "burst limits for Access API methods e.g. Ping=100,GetTransaction=100 etc.")
+		flags.BoolVar(&fnb.Unstaked, "unstaked", false, "whether this node is an unstaked access node")
```
Nit: I wonder if it is more natural to have a new param `staked` that defaults to true, instead of `unstaked` that defaults to false?

smnzhu replied:
Should we add some validation for these new fields?

It seems like at the very least, we should probably check that if `staked` is false (or `unstaked` is true), then `staked-access-node-addr` should not be empty right?

smnzhu replied:
Should we add some validation for these new fields?

It seems like at the very least, we should probably check that if `staked` is false (or `unstaked` is true), then `staked-access-node-addr` should not be empty right?

vishalchangrani replied:
yeah agreed..lemme make those changes (and some more)

vishalchangrani replied:
> Nit: I wonder if it is more natural to have a new param `staked` that defaults to true, instead of `unstaked` that defaults to false?

changed the flag to be `staked` now. It definitely reads better as you suggested

smnzhu commented on cmd/scaffold.go:
```diff
@@ -204,6 +205,16 @@ func (fnb *FlowNodeBuilder) enqueueNetworkInit() {
 
+		if fnb.Unstaked {
+			fnb.MsgValidators = []network.MessageValidator{
+				// filter out messages sent by this node itself
+				validator.NewSenderValidator(fnb.Me.NodeID()),
+				// but retain all the 1-k messages even if they are not intended for this node
```
Can you explain this part a little more? Why do we filter out messages sent by this node itself, and what is the significance of 1-k?

vishalchangrani replied:
sure..currently, when you send a 1-k message (in other words, publish a message on a topic), the network layers seems to also loopback the message to the sender if the sender has also subscribed to the topic. This is not desired and hence a validator was added to filter out such messages which have the Origin field in the message equal to the node ID of the node (loop back messages).

Also, when you send a 1-k message, the sender has to specify a list of TargetIDs which indicate the nodes for which the message is intended. This is generally a subset of nodes subscribed on the topic. e.g. If a collection node sends a tx to collection node 1, 2 and 3, it will be received by ALL collection nodes in the system since they are all subscribed to the send-transaction channel. The target validator makes sure that only the intended targets receive the message and the other nodes just drop it.

smnzhu replied:
I guess my question is, why does the behavior of the staked vs unstaked node need to be different?

huitseeker replied:
@vishalchangrani does it simplify the management of the current unstaked access node set if unstaked receivers can pick up messages destined for an obsolete ID?

vishalchangrani commented on cmd/access/main.go:
```diff
@@ -103,6 +108,35 @@ func main() {
+				Role:          flow.RoleAccess, // unstaked node can only run as an access node
+				Address:       node.BaseConfig.BindAddr,
+			}
+			node.Me, err = local.New(self, nil)
```
setting `node.Me` here makes sure that the FlowNodeBuilder does not attemps to initialize the node id from the protocol state.

huitseeker replied:
That's actually a detail I'd love to read as a code comment.

vishalchangrani commented on cmd/access/main.go:
```diff
@@ -103,6 +108,43 @@ func main() {
+			if staked {
+				return
+			}
+			if strings.TrimSpace(stakedAccessNodeIDHex) == "" {
```
yet to add code to actually use the staked access node id

vishalchangrani commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,195 @@
+
+// AccessNodeBuilder is initializes and runs an Access node either as a staked node or an unstaked node
+// It is composed of the FlowNodeBuilder
+type AccessNodeBuilder struct {
```
Given the way an unstaked access node (and a staked access node which is supporting unstaked access nodes) need to be bootstrapped, a custom NodeBuilder is needed. The FlowNodeBuilder does most of the initialization but it needs be customized to bring up the unstaked network and skip certain check for the unstaked access node.

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -672,6 +680,11 @@ func (fnb *FlowNodeBuilder) Component(name string, f func(*FlowNodeBuilder) (mod
 	return fnb
 }
 
+func (fnb *FlowNodeBuilder) PreInit(f func(node *FlowNodeBuilder)) *FlowNodeBuilder {
```
These functions help customize the initialization. In the case of the unstaked access node, it lets setting the `local` for the node so that it is attempted to be initialized later from the protocol state.

huitseeker replied:
We may need a code comment that clarifies the two code paths here:
- one where the initialization runs from the protocol state, i.e. the normal path
- one where that info is not in the protocol state accessible to the node, and therefore needs to be set through other means (the unstaked node path)

vishalchangrani commented on cmd/scaffold.go:
```diff
@@ -682,23 +695,32 @@ func FlowNode(role string) *FlowNodeBuilder {
+}
 
-	builder.enqueueNetworkInit()
+func (fnb *FlowNodeBuilder) Initialize() *FlowNodeBuilder {
```
broken the constructor `FlowNode` into `FlowNode` and `Initialize`. For the access node, the initialization is different and breaking down the function helps that custom initialization.

vishalchangrani commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,195 @@
+
+func (anb *AccessNodeBuilder) EnqueueUnstakedNetworkInit() {
+	anb.Component("unstaked network", func(node *cmd.FlowNodeBuilder) (module.ReadyDoneAware, error) {
+		codec := jsoncodec.NewCodec()
```
creating the new unstaked network to be used between the unstaked access nodes and the staked access node.

smnzhu commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,195 @@
+	*cmd.FlowNodeBuilder
+	staked                bool
+	stakedAccessNodeIDHex string
+	unstakedBindAddr      string
```
Nit: `unstakedNetworkBindAddr` would be more clear imo

vishalchangrani commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,195 @@
+
+		// for an unstaked node, use message sender validator but not target validator since the staked AN will
+		// be broadcasting messages to ALL unstaked ANs without knowing their target IDs
+		msgValidators := []network.MessageValidator{
+			// filter out messages sent by this node itself
+			validator.NewSenderValidator(node.Me.NodeID()),
+			// but retain all the 1-k messages even if they are not intended for this node
+		}
```
agreed..will change.

vishalchangrani commented on cmd/collection/main.go:
```diff
@@ -126,6 +126,7 @@ func main() {
 			flags.DurationVar(&blockRateDelay, "block-rate-delay", 250*time.Millisecond,
 				"the delay to broadcast block proposal in order to control block production rate")
 		}).
+		Initialize().
```
initializing all node types after all the flags have been defined. 
The access node needs the extra flag to determine how to initialize itself.

smnzhu commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+		// creates network instance
+		net, err := p2p.NewNetwork(node.Logger,
+			codec,
+			participants,
```
Is this correct? `participants` is the list of staked nodes, right? Does it make sense to use these as the participants for the unstaked network?

huitseeker replied:
👍 
More exactly: are we sure that, even if it wanted, the unstaked node cannot access staked network information (see my comment on the PR on having a != bootstrap config)

smnzhu commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+		}
+
+		// creates topology, topology manager, and subscription managers
+		//
+		// topology
+		// subscription manager
```
Clean up comments before merge?

smnzhu commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+			participants,
+			node.Me,
+			anb.unstakedMiddleware,
+			10e6,
```
nit: maybe create a constant for this?

vishalchangrani replied:
took care of this in https://github.com/onflow/flow-go/pull/983

huitseeker commented on cmd/access/main.go:
```diff
@@ -325,8 +329,11 @@ func main() {
-		Component("ping engine", func(node *cmd.FlowNodeBuilder) (module.ReadyDoneAware, error) {
+		})
+
+	// the ping engine is only needed for the staked access node
```
I'm not sure this isn't needed: cf, my comment on unstaked node fault detection. We might need a one-directional ping engine (from the staked AN to the unstaked AN) though.

vishalchangrani replied:
The ping will be useful but I see that as an additional feature (staked AN being able to track liveness on unstaked AN) and thought I could tackle that separately as a different issue. The current ping reports Node ID and RTT time of all staked nodes and it will have to be changed to report a different metric for unstaked nodes to keep them separate on Grafana.

huitseeker reviewed (commented):
I have a couple questions on concept for the setup of both the unstaked node and that staked access node that may connect to them:
- since unstaked access nodes don't suffer from leaving the network, how do we make sure the access node periodically trims a connection to them if they are down?
- could we bootstrap the unstaked access node from a separate fileset, even in our localnet? And could we have the feature that Init() crashes on unstaked nodes if they have a set of participants that includes anything with a staking key but their upstream AN? Could we even not put the upstream AN's staking key in their config, so we're sure they don't get more info than needed? That way we detect if we have unduly given staked network info to unstaked nodes
- Would a crasher in the unstaked network's libp2p Host bring down the staked access node?  if so, how do we track that risk?

On the code more specifically:
- 👍  on the breakdown of the FlowNode config, Init / PreInit makes sense,
- with that said, we're opening a very wide door in making baseFlags, enqueueNetworkInit, etc public. Might we leave at least some of them private, hide them behind an interface, have both nodes (staked & unstaked) implement that and run one single public method run through the initialization sequence on the interface?

smnzhu commented:
> with that said, we're opening a very wide door in making baseFlags, enqueueNetworkInit, etc public. Might we leave at least some of them private, hide them behind an interface, have both nodes (staked & unstaked) implement that and run one single public method run through the initialization sequence on the interface?

I think this makes sense, but if i recall correctly the reason we initially decided to make them public is because we wanted to share the code between unstaked and staked node, but the `AccessNodeBuilder` is in different package from `scaffold.go`, and therefore the only way for them to share the code is to make the code public?

huitseeker commented:
I may be missing something, but would it be possible to have an interface [at the package level of scaffold](https://github.com/golang/go/wiki/CodeReviewComments#interfaces), and the implementation of the `AccessNodeBuilder` would embed a `FlowNodeBuilder`, with the proper indirections / reimplementations on top?

vishalchangrani commented:
> I may be missing something, but would it be possible to have an interface [at the package level of scaffold](https://github.com/golang/go/wiki/CodeReviewComments#interfaces), and the implementation of the `AccessNodeBuilder` would embed a `FlowNodeBuilder`, with the proper indirections / reimplementations on top?

I think that will be possible. Lemme work on that. I didn't realize I would need so many changes on the base class until after I finished it.

vishalchangrani commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+	anb.Component("unstaked network", func(node *cmd.FlowNodeBuilder) (module.ReadyDoneAware, error) {
+		codec := jsoncodec.NewCodec()
+
+		// setup the Ping provider to return the software version and the finalized block height
```
good catch!

vishalchangrani commented:
Refactored the code to address this [comment](https://github.com/onflow/flow-go/pull/911#issuecomment-879261062).

Now, the structure is as follows:
1. A new interface [NodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/node_builder.go) defines all methods needed to bootstrap a Flow node.
2. Existing [FlowNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/scaffold.go#L97) and the new [AccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/access_node_builder.go#L29) implement this interface.
3. A new interface [AccessNodeBuilder](https://github.com/onflow/flow-go/blob/vishal/unstaked_an/cmd/access/access_node_builder.go#L21) which extends the NodeBuilder interface and adds Access Node specific functions.
4. Separate `StakedAccessNodeBuilder` and `UnstakedAccessNodeBuilder` (which are composed of `AccessNodeBuilder`) to take care of bringing up Staked and Unstaked AN resp.
5. Changes to all existing startup files (`main.go`) to use the `NodeBuilder` functions instead of directly using the implementation.

vishalchangrani commented:
> I have a couple questions on concept for the setup of both the unstaked node and that staked access node that may connect to them:
> 
> * since unstaked access nodes don't suffer from leaving the network, how do we make sure the access node periodically trims a connection to them if they are down?
> * could we bootstrap the unstaked access node from a separate fileset, even in our localnet? And could we have the feature that Init() crashes on unstaked nodes if they have a set of participants that includes anything with a staking key but their upstream AN? Could we even not put the upstream AN's staking key in their config, so we're sure they don't get more info than needed? That way we detect if we have unduly given staked network info to unstaked nodes.
> * that configuration file could be read by the staked AN to bootstrap its unstaked network as well.
> * Would a crasher in the unstaked network's libp2p Host bring down the staked access node?  if so, how do we track that risk?
> 
> On the code more specifically:
> 
> * 👍  on the breakdown of the FlowNode config, Init / PreInit makes sense,
> * with that said, we're opening a very wide door in making baseFlags, enqueueNetworkInit, etc public. Might we leave at least some of them private, hide them behind an interface, have both nodes (staked & unstaked) implement that and run one single public method run through the initialization sequence on the interface?

@huitseeker - To address the issue of not letting the unstaked AN know about the identities of the staked node at all, I have created this new issue [The unstaked Access Node should bootstrap without the identity list of the staked network](https://github.com/dapperlabs/flow-go/issues/5674)

vishalchangrani commented:
@jordanschalm @huitseeker @smnzhu - To help make the code review a little easier, I moved all the bootstrap changes to scaffold into its own separate PR here - https://github.com/onflow/flow-go/pull/983. This new PR is based off of this branch (`unstaked_an`). I reverted `unstaked_an` to what it was before those changes.

jordanschalm commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+	unstakedMiddleware      *p2p.Middleware
+}
+
+func FlowAccessNode() *AccessNodeBuilder {
```
We could move this and the existing `FlowNodeBuilder` to a `scaffold` package together to avoid all the newly public methods in `FlowNodeBuilder` and keep the scaffolding logic together.

vishalchangrani replied:
Lemme do this in the other [PR](https://github.com/onflow/flow-go/pull/983) since that one is primarily for such scaffold related changes.

vishalchangrani replied:
Actually, lemme do that as different PR all together once I have [983](https://github.com/onflow/flow-go/pull/983) merged into this one otherwise it will just become even more difficult to review.

vishalchangrani replied:
created [this](https://github.com/dapperlabs/flow-go/issues/5688) issue to address this comment

jordanschalm commented on cmd/access/access_node_builder.go:
```diff
@@ -0,0 +1,200 @@
+// It is composed of the FlowNodeBuilder
+type AccessNodeBuilder struct {
+	*cmd.FlowNodeBuilder
+	staked                  bool
```
I wonder if there is a different term than "unstaked" that we can use for this new mode of operation for the Access Node. Being unstaked doesn't really describe the new mode of operation very well IMO, and the notions of stake and weight are already a bit overloaded and frankly confusing in parts in the protocol code (see [here](https://github.com/onflow/flow-go/blob/c59b5e4de1b59766d352209dbc856068c03f9225/model/flow/identity.go#L30-L39) for an example). 

For example, in the [`NodeConfig`](https://github.com/onflow/flow-go/pull/911/files#diff-7540be57ca1fb2714cfce630ad0eba7bfc97362d8e30b79ac2551dc68792e75bR252) we have a field called `Stake`, which really means `Weight` and we have a field called `Unstaked` which only applies to access nodes, doesn't have much to do with the field called `Stake`, and determines whether an AN runs in this new external network or as part of the core Flow network.

Some ideas: External AN / network, Proxy AN / network, Adjunct AN / network

vishalchangrani replied:
Public network? (versus our current private network)

vishalchangrani replied:
tbh - i like the word `unstaked network` it succinctly describes this network as a network of unstaked actors. Any other name seems to be requiring more explanation since our _flow network_ is public and external.

jordanschalm replied:
Maybe in places where the new field / name is near an existing stake/weight reference we can expand the name to differentiate or add some comments to differentiate

smnzhu commented on cmd/access/staked_access_node_builder.go:
```diff
@@ -0,0 +1,101 @@
+
+// SupportUnstakedNodes returns True if this a staked Access Node which also participates in the unstaked network,
+// False otherwise
+func (builder *StakedAccessNodeBuilder) supportUnstakedNodes() bool {
```
I think we should put this function on the parent `FlowAccessNodeBuilder`, and rename to something like `ParticipatesInUnstakedNetwork`.

This is because the `main.go` will need to use this function to determine whether to create a splitter engine and relay engine for a staked access node.

vishalchangrani replied:
done

smnzhu commented on cmd/node_builder.go:
```diff
@@ -0,0 +1,130 @@
+	Storage           Storage
+	ProtocolEvents    *events.Distributor
+	State             protocol.State
+	Middleware        *p2p.Middleware
+	Network           *p2p.Network
```
Let's replace these with the interfaces `network.Middleware` and `module.Network`

vishalchangrani commented:
@jordanschalm @huitseeker @smnzhu - thanks for the diligent review on this PR so far! ❤️ 
I have merged in the other [PR](https://github.com/onflow/flow-go/pull/983) that was reviewed by y'all into this one.
Please can you review and approve this PR as well? I believe I have addressed all the outstanding comments.
thanks 🙇

smnzhu commented on cmd/access/access_node_builder.go:
```diff
@@ -53,20 +53,32 @@ func FlowAccessNode() *FlowAccessNodeBuilder {
-func (anb *FlowAccessNodeBuilder) parseFlags() {
+// supportUnstakedNodes returns True if this a staked Access Node which also participates in the unstaked network,
+// False otherwise
+func (builder *FlowAccessNodeBuilder) supportUnstakedNodes() bool {
```
What I meant was to add it to the `AccessNodeBuilder` interface, not only the `FlowAccessNodeBuilder` implementation. It will also need to be public. See [this](https://github.com/onflow/flow-go/commit/59ae3f825929ceef58390657367ece1a0ef71aa2#r53749799) for example.

We would rename the method as `ParticipatesInUnstakedNetwork`, and it would always return true for unstaked nodes. 

Of course, instead of doing this I could just do a type assertion [here](https://github.com/onflow/flow-go/commit/59ae3f825929ceef58390657367ece1a0ef71aa2#r53750040), but I think this way is a little bit cleaner.

vishalchangrani replied:
oh shoot! I forgot to push the commit with that change. Yes, 100% agree that method should be on the interface.

huitseeker reviewed (approved):
Superb work!

jordanschalm reviewed (approved):
Nice work 🎸

vishalchangrani commented:
> Nice work 🎸

thanks for all the amazing feedback..really appreciate it!

jordanschalm commented on cmd/access/main.go:
```diff
@@ -125,7 +145,7 @@ func main() {
 			return err
 		}).
-		Module("collection node client", func(node *cmd.FlowNodeBuilder) error {
+		Module("collection node client", func(builder cmd.NodeBuilder, node *cmd.NodeConfig) error {
```
@vishalchangrani Why are we passing in both `NodeBuilder` and `NodeConfig`? It looks like the builder struct embeds the config.
